### PR TITLE
credential ramfs mount order fixes

### PR DESCRIPTION
**Pull request by poettering**
A comprehensive fix for #25527

### CODE REVIEW

File: src/core/unit.c [hunk 1]
```diff
@@ -1349,6 +1349,27 @@ int unit_add_exec_dependencies(Unit *u, ExecContext *c) {
         if (r < 0)
                 return r;
 
+        if (exec_context_has_credentials(c) && u->manager->prefix[EXEC_DIRECTORY_RUNTIME]) {
+                _cleanup_free_ char *m = NULL;
+                const char *p;
+
+                /* Let's make sure the credentials directory of this service is unmounted *after* the service
+                 * itself shuts down. This only matters if mount namespacing is not used for the service, and
+                 * hence the credenitals mount appears on the host. */
```
Comment (daandemeyer) [inline src/core/unit.c]: Typo

File: src/core/unit.c [hunk 2]
```diff
@@ -1349,6 +1349,27 @@ int unit_add_exec_dependencies(Unit *u, ExecContext *c) {
...
+
+                p = path_join(u->manager->prefix[EXEC_DIRECTORY_RUNTIME], "credentials", u->id);
+                if (!p)
+                        return -ENOMEM;
+
+                r = unit_name_from_path(p, ".mount", &m);
+                if (r < 0)
+                        return r;
+
+                r = unit_add_dependency_by_name(u, UNIT_AFTER, m, true, UNIT_DEPENDENCY_FILE);
```
Comment (daandemeyer) [inline src/core/unit.c]: Missing `/* variable = */` for the `true`

File: src/core/mount.c [hunk 3]
```diff
@@ -475,6 +475,26 @@ static bool mount_is_extrinsic(Unit *u) {
         return false;
 }
 
+static bool mount_is_credentials(Mount *m) {
+        const char *e;
+
+        assert(m);
+
+        /* Returns true if this is a credentials mount. We don't want automatic dependencies on credentials,
+         * since they are managed by us for event he earliest services, and we never want anythong to be
```
Comment (daandemeyer) [inline src/core/mount.c]: Typos

File: src/core/mount.c [hunk 4]
```diff
@@ -475,6 +475,26 @@ static bool mount_is_extrinsic(Unit *u) {
...
+
+        e = path_startswith(m->where, UNIT(m)->manager->prefix[EXEC_DIRECTORY_RUNTIME]);
+        if (!e)
+                return false;
+
+        e = path_startswith(e, "credentials");
+        if (!e)
+                return false;
+
+        return !isempty(e);
```
Comment (daandemeyer) [inline src/core/mount.c]: `return !isempty(path_startswith(e, "credentials"));` ?

File: man/sd_event_source_set_ratelimit.xml [hunk 5]
```diff
@@ -85,10 +90,14 @@
     is currently affected by rate limiting, i.e. it has recently hit the rate limit and is currently
     temporarily disabled due to that.</para>
 
-    <para><function>sd_event_source_set_ratelimit_expire_callback</function> may be used to set a callback
+    <para><function>sd_event_source_set_ratelimit_expire_callback()</function> may be used to set a callback
     function that is invoked every time the event source leaves rate limited state. Note that function is
     called in the same event loop iteration in which state transition occurred.</para>
 
+    <para><function>sd_event_source_leave_ratelimit()</function> may be used to immediately reenable am event
```
Comment (daandemeyer) [inline man/sd_event_source_set_ratelimit.xml]: Typo

File: man/sd_event_source_set_ratelimit.xml [hunk 6]
```diff
@@ -98,10 +107,12 @@
 
     <para>On success, <function>sd_event_source_set_ratelimit()</function>,
     <function>sd_event_source_set_ratelimit_expire_callback</function> and
-    <function>sd_event_source_get_ratelimit()</function> return a non-negative integer. On failure, they return
-    a negative errno-style error code. <function>sd_event_source_is_ratelimited</function> returns zero if rate
-    limiting is currently not in effect and greater than zero if it is in effect; it returns a negative
-    errno-style error code on failure.</para>
+    <function>sd_event_source_get_ratelimit()</function> return a non-negative integer. On failure, they
+    return a negative errno-style error code. <function>sd_event_source_is_ratelimited</function> returns
```
Comment (daandemeyer) [inline man/sd_event_source_set_ratelimit.xml]: Can add () to sd_event_source_is_ratelimited here as well

File: man/sd_event_source_set_ratelimit.xml [hunk 7]
```diff
@@ -98,10 +107,12 @@
...
+    zero if rate limiting is currently not in effect and greater than zero if it is in effect; it returns a
+    negative errno-style error code on failure. <function>sd_event_source_leave_ratelimit</function> returns
```
Comment (daandemeyer) [inline man/sd_event_source_set_ratelimit.xml]: Ditto

File: src/core/mount.c [hunk 8]
```diff
@@ -1569,6 +1597,10 @@ static void mount_sigchld_event(Unit *u, pid_t pid, int code, int status) {
                 mount_enter_dead_or_mounted(m, f);
                 break;
 
+        case MOUNT_UNMOUNTING_CATCHUP:
+                /* Didn't expect a process to die, hence let's ignore it */
```
Comment (daandemeyer) [inline src/core/mount.c]: Maybe let's debug log here

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
I'm confused about the last commit, what is unmounting the credentials mount if not us?

**Comment by poettering** [comment]
> I'm confused about the last commit, what is unmounting the credentials mount if not us?

It is us, but the nasty part is that /proc/self/mountinfo notifications are async, so if we stop a service (and while doing so get rid of the credentials mount point of it), then it will take a while until the notification reaches us and we actually scan the table again. In particular as we nowadays ratelimit notifications on the table, since it's so inefficient. And as I learnt the ratelimiting is actually quite regularly hit during shutdown, where a flurry of umount events are genreated.

That said, there are other daemons that unmount stuff when stopping where the race also exists, but might simply be harder to trigger: if during service shutdown these services remove some mount then they might collide with us doing the same. After all, we have the rule to unmount everything mounted automatically for you during shutdown.

In the long run we should also start making us of this when it becomes available: https://github.com/util-linux/util-linux/issues/2132 With that we can make issues like this go away entirely from our side of things at least.

**Comment by DaanDeMeyer** [comment]
Can you put this in some form in the commit message?

**Review by yuwata — DISMISSED** [review]
LGTM.

**Comment by mrc0mmand** [comment]
LSan is still sad:

```
==1==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 384 byte(s) in 6 object(s) allocated from:
    #0 0x7f525070ea0a  (/usr/lib/clang/15.0.7/lib/linux/libclang_rt.asan-x86_64.so+0x10ea0a) (BuildId: faf0f22f6ed3e575962cd50bcfba6f2e17878558)
    #1 0x7f524f06953b in path_extend_internal /systemd-meson-build/../build/src/basic/path-util.c:558:14
    #2 0x7f5250149013 in unit_add_exec_dependencies /systemd-meson-build/../build/src/core/unit.c:1360:21
    #3 0x7f5250005e84 in service_add_extras /systemd-meson-build/../build/src/core/service.c:777:13
    #4 0x7f5250005e84 in service_load /systemd-meson-build/../build/src/core/service.c:822:13
    #5 0x7f52501515fa in unit_load /systemd-meson-build/../build/src/core/unit.c:1648:13
    #6 0x7f524ff755d2 in manager_dispatch_load_queue /systemd-meson-build/../build/src/core/manager.c:2137:17
    #7 0x7f524ff749a1 in manager_load_unit /systemd-meson-build/../build/src/core/manager.c:2268:9
    #8 0x7f525014a234 in unit_add_dependency_by_name /systemd-meson-build/../build/src/core/unit.c:3232:13
    #9 0x7f524ffaeef6 in mount_add_default_ordering_dependencies /systemd-meson-build/../build/src/core/mount.c:530:21
    #10 0x7f524ffaeef6 in mount_add_default_dependencies /systemd-meson-build/../build/src/core/mount.c:605:13
    #11 0x7f524ffaeef6 in mount_add_non_exec_dependencies /systemd-meson-build/../build/src/core/mount.c:672:13
    #12 0x7f524ffbee62 in mount_setup_new_unit /systemd-meson-build/../build/src/core/mount.c:1730:13
    #13 0x7f524ffbee62 in mount_setup_unit /systemd-meson-build/../build/src/core/mount.c:1854:21
    #14 0x7f524ffbee62 in mount_load_proc_self_mountinfo /systemd-meson-build/../build/src/core/mount.c:1899:24
    #15 0x7f524ffacdca in mount_enumerate /systemd-meson-build/../build/src/core/mount.c:2056:13
    #16 0x7f524ff6f98c in manager_enumerate /systemd-meson-build/../build/src/core/manager.c:1666:25
    #17 0x7f524ff6dbef in manager_startup /systemd-meson-build/../build/src/core/manager.c:1872:17
    #18 0x56494f7ec2c3 in main /systemd-meson-build/../build/src/core/main.c:3053:13
    #19 0x7f524e63c78f  (/usr/lib/libc.so.6+0x2378f) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)

Direct leak of 384 byte(s) in 6 object(s) allocated from:
    #0 0x7f525070ea0a  (/usr/lib/clang/15.0.7/lib/linux/libclang_rt.asan-x86_64.so+0x10ea0a) (BuildId: faf0f22f6ed3e575962cd50bcfba6f2e17878558)
    #1 0x7f524f06953b in path_extend_internal /systemd-meson-build/../build/src/basic/path-util.c:558:14
    #2 0x7f5250149013 in unit_add_exec_dependencies /systemd-meson-build/../build/src/core/unit.c:1360:21
    #3 0x7f5250005e84 in service_add_extras /systemd-meson-build/../build/src/core/service.c:777:13
    #4 0x7f5250005e84 in service_load /systemd-meson-build/../build/src/core/service.c:822:13
    #5 0x7f52501515fa in unit_load /systemd-meson-build/../build/src/core/unit.c:1648:13
    #6 0x7f524ff755d2 in manager_dispatch_load_queue /systemd-meson-build/../build/src/core/manager.c:2137:17
    #7 0x7f524ff749a1 in manager_load_unit /systemd-meson-build/../build/src/core/manager.c:2268:9
    #8 0x7f525014a234 in unit_add_dependency_by_name /systemd-meson-build/../build/src/core/unit.c:3232:13
    #9 0x7f524ffaeef6 in mount_add_default_ordering_dependencies /systemd-meson-build/../build/src/core/mount.c:530:21
    #10 0x7f524ffaeef6 in mount_add_default_dependencies /systemd-meson-build/../build/src/core/mount.c:605:13
    #11 0x7f524ffaeef6 in mount_add_non_exec_dependencies /systemd-meson-build/../build/src/core/mount.c:672:13
    #12 0x7f524ffbee62 in mount_setup_new_unit /systemd-meson-build/../build/src/core/mount.c:1730:13
    #13 0x7f524ffbee62 in mount_setup_unit /systemd-meson-build/../build/src/core/mount.c:1854:21
    #14 0x7f524ffbee62 in mount_load_proc_self_mountinfo /systemd-meson-build/../build/src/core/mount.c:1899:24
    #15 0x7f524ffacdca in mount_enumerate /systemd-meson-build/../build/src/core/mount.c:2056:13
    #16 0x7f524ff6f98c in manager_enumerate /systemd-meson-build/../build/src/core/manager.c:1666:25
    #17 0x7f524ff87073 in manager_reload /systemd-meson-build/../build/src/core/manager.c:3481:9
    #18 0x56494f7f79df in invoke_main_loop /systemd-meson-build/../build/src/core/main.c:1995:29
    #19 0x56494f7ec4a1 in main /systemd-meson-build/../build/src/core/main.c:3081:13
    #20 0x7f524e63c78f  (/usr/lib/libc.so.6+0x2378f) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)

Direct leak of 64 byte(s) in 1 object(s) allocated from:
    #0 0x7f525070ea0a  (/usr/lib/clang/15.0.7/lib/linux/libclang_rt.asan-x86_64.so+0x10ea0a) (BuildId: faf0f22f6ed3e575962cd50bcfba6f2e17878558)
    #1 0x7f524f06953b in path_extend_internal /systemd-meson-build/../build/src/basic/path-util.c:558:14
    #2 0x7f5250149013 in unit_add_exec_dependencies /systemd-meson-build/../build/src/core/unit.c:1360:21
    #3 0x7f5250005e84 in service_add_extras /systemd-meson-build/../build/src/core/service.c:777:13
    #4 0x7f5250005e84 in service_load /systemd-meson-build/../build/src/core/service.c:822:13
    #5 0x7f52501515fa in unit_load /systemd-meson-build/../build/src/core/unit.c:1648:13
    #6 0x7f524ff755d2 in manager_dispatch_load_queue /systemd-meson-build/../build/src/core/manager.c:2137:17
    #7 0x7f524ff749a1 in manager_load_unit /systemd-meson-build/../build/src/core/manager.c:2268:9
    #8 0x7f524ff63992 in manager_deserialize_one_unit /systemd-meson-build/../build/src/core/manager-serialize.c:211:13
    #9 0x7f524ff63992 in manager_deserialize_units /systemd-meson-build/../build/src/core/manager-serialize.c:243:21
    #10 0x7f524ff63992 in manager_deserialize /systemd-meson-build/../build/src/core/manager-serialize.c:573:16
    #11 0x7f524ff87095 in manager_reload /systemd-meson-build/../build/src/core/manager.c:3484:13
    #12 0x56494f7f79df in invoke_main_loop /systemd-meson-build/../build/src/core/main.c:1995:29
    #13 0x56494f7ec4a1 in main /systemd-meson-build/../build/src/core/main.c:3081:13
    #14 0x7f524e63c78f  (/usr/lib/libc.so.6+0x2378f) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)

Direct leak of 64 byte(s) in 1 object(s) allocated from:
    #0 0x7f525070ea0a  (/usr/lib/clang/15.0.7/lib/linux/libclang_rt.asan-x86_64.so+0x10ea0a) (BuildId: faf0f22f6ed3e575962cd50bcfba6f2e17878558)
    #1 0x7f524f06953b in path_extend_internal /systemd-meson-build/../build/src/basic/path-util.c:558:14
    #2 0x7f5250149013 in unit_add_exec_dependencies /systemd-meson-build/../build/src/core/unit.c:1360:21
    #3 0x7f5250005e84 in service_add_extras /systemd-meson-build/../build/src/core/service.c:777:13
    #4 0x7f5250005e84 in service_load /systemd-meson-build/../build/src/core/service.c:822:13
    #5 0x7f52501515fa in unit_load /systemd-meson-build/../build/src/core/unit.c:1648:13
    #6 0x7f524ff755d2 in manager_dispatch_load_queue /systemd-meson-build/../build/src/core/manager.c:2137:17
    #7 0x7f524ff749a1 in manager_load_unit /systemd-meson-build/../build/src/core/manager.c:2268:9
    #8 0x7f524ff771f9 in manager_load_startable_unit_or_warn /systemd-meson-build/../build/src/core/manager.c:2285:13
    #9 0x56494f7f69dc in do_queue_default_job /systemd-meson-build/../build/src/core/main.c:2314:13
    #10 0x56494f7ec34d in main /systemd-meson-build/../build/src/core/main.c:3064:21
    #11 0x7f524e63c78f  (/usr/lib/libc.so.6+0x2378f) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)

SUMMARY: AddressSanitizer: 896 byte(s) leaked in 14 allocation(s).
```

**Comment by poettering** [comment]
ouch. that was a dumb one.

**Comment by yuwata** [comment]
Hmm... I am confused again.

@poettering 
- First of all, why we try to setup the credential directory as a ramfs mount?
- Of course, the ramfs mount is optional. But, then why the ramfs is mounted before creating a mount namespace, rather than after?

File: src/core/mount.c [hunk 9]
```diff
@@ -1073,11 +1074,37 @@ static void mount_enter_unmounting(Mount *m) {
 
         assert(m);
 
+        r = path_is_mount_point(m->where, NULL, 0);
```
Comment (yuwata) [inline src/core/mount.c:1074]: This is still racy. The mount point may be unmounted after `path_is_mount_point()` succeeds and before invoking `umount` command.

File: src/libsystemd/sd-event/sd-event.c [hunk 10]
```diff
@@ -5153,6 +5153,27 @@ _public_ int sd_event_source_is_ratelimited(sd_event_source *s) {
         return s->ratelimited;
 }
 
+_public_ int sd_event_source_leave_ratelimit(sd_event_source *s) {
```
Comment (yuwata) [inline src/libsystemd/sd-event/sd-event.c:5196]: Maybe better to add a brief test about this?

**Comment by poettering** [comment]
> First of all, why we try to setup the credential directory as a ramfs mount?

ramfs (as supposed to tmpfs) is not backed by swap, i.e. locked into memory. Since credentials are supposed to be useful for secrets that's preferable.

> Of course, the ramfs mount is optional. But, then why the ramfs is mounted before creating a mount namespace, rather than after?

It's not. When a service has namespacing configured we'll mount the ramfs inside the namespace, nothing will be visible on the host. However, it was important to me to ensure credentials also work without namespaces, and without privs (in containers and --user) hence we gracefully degrade: install the mounts on the host, or not at all if we cannot.

> How about to setup the credential directory as a ramfs mount only when the service enables mount namespace?

mounts are ultimately cheap if they are ramfs/tmpfs hence I figured we might as well make things secure

File: src/core/mount.c (hunk 9 above)
Comment (poettering) [reply src/core/mount.c:1074]: see comment in commit msg. Yes, the race exists, this PR only addresses it if PID 1 itself is the culprit: we won't race against our own mount attempts anymore. But yes, in the long run we want https://github.com/util-linux/util-linux/issues/2132

**Comment by yuwata** [comment]
> When a service has namespacing configured we'll mount the ramfs inside the namespace, nothing will be visible on the host.

It seems not. `setup_credentials()` is called before `apply_mount_namespace()`:
https://github.com/systemd/systemd/blob/f461a28da70320473c7466bcdc34fa601da0877b/src/core/execute.c#L4885
https://github.com/systemd/systemd/blob/f461a28da70320473c7466bcdc34fa601da0877b/src/core/execute.c#L5091

**Comment by poettering** [comment]
or in other words: I think it would make sense to rework the logic eventually, based on the new mount api: allocate the ramfs and populate it via the new mount api, and then store it somewhere in the unit while it is active, and attach it to each ExecXYZ= line we activate.

File: src/core/mount.c [hunk 11]
```diff
@@ -1613,6 +1645,7 @@ static int mount_dispatch_timer(sd_event_source *source, usec_t usec, void *user
                 break;
 
         case MOUNT_UNMOUNTING:
+        case MOUNT_UNMOUNTING_CATCHUP:
                 log_unit_warning(UNIT(m), "Unmounting timed out. Terminating.");
                 mount_enter_signal(m, MOUNT_UNMOUNTING_SIGTERM, MOUNT_FAILURE_TIMEOUT);
```
Comment (yuwata) [inline src/core/mount.c]: When MOUNT_UNMOUNTING_CATCHUP, calling `mount_enter_signal()` without invoking `umount` command sounds spurious for me. Also, the log message `"Terminating"` sounds spurious in this case.
How about to call `mount_enter_dead_or_mounted(m, MOUNT_FAILURE_TIMEOUT);` instead?

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by yuwata** [comment]
> or in other words: I think it would make sense to rework the logic eventually, based on the new mount api: allocate the ramfs and populate it via the new mount api, and then store it somewhere in the unit while it is active, and attach it to each ExecXYZ= line we activate.

Sounds good for me.

File: src/libsystemd/sd-event/sd-event.c (hunk 10 above)
Comment (yuwata) [reply src/libsystemd/sd-event/sd-event.c:5196]: Still I'd like to have a simple test case for this.

Comment (poettering) [reply src/libsystemd/sd-event/sd-event.c:5196]: yupp, working on it

File: src/core/mount.c (hunk 11 above)
Comment (poettering) [reply src/core/mount.c]: indeed

**Comment by poettering** [comment]
Force pushed a new version adding the requested test and with the suggested mount unit timeout state change. ptal!

File: src/libsystemd/libsystemd.sym [hunk 12]
```diff
@@ -824,4 +824,5 @@ global:
         sd_event_source_set_memory_pressure_period;
         sd_event_trim_memory;
         sd_pid_notify_barrier;
+        sd_event_source_leave_ratelimit;
```
Comment (daandemeyer) [inline src/libsystemd/libsystemd.sym:827]: Uhhh this needs to be moved to v254

**Review by DaanDeMeyer — APPROVED** [review]
Looks good, but needs a version fix in the sym file

**Review by yuwata — APPROVED** [review]
LGTM.

(Still I'd like to make PID1 not trigger mount event on the main mount namespace if possible (that is, when a new mount namespace will be created for a unit). But, that's out of scope of this PR. I think.)

File: src/libsystemd/libsystemd.sym (hunk 12 above)
Comment (poettering) [reply src/libsystemd/libsystemd.sym:827]: it *is* in the 254 section. (this is confusing, indeed, but open the context of this patch and you'll see we define this in the 254 section and then chain it up with the 253 section)

**Comment by poettering** [comment]
thanks! let's merge, given that the one point raised is a misunderstanding.

**Pull request merged**